### PR TITLE
SC-1960 Migration of data for user_lookup

### PR DESCRIPTION
**Pull request by sknirmalkar89**
Migrating user_lookup table for email, phone, username and externalId

User Table records:41
User Org Table records:39
State Users Table Records:754622

#Total records which contains alteast one identifiers email, phone, username
user data Count : 41

#Valid State Users
State Users:754617

#valid userid and orgId Mapping
User Org Mapping count:32

Total Duplicate Emails: 35
Total Duplicate phone: 8
Total Duplicate username: 0

user_lookup count post migration: 754479

### CODE REVIEW

File: etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    val userOrgSchema = Encoders.product[UserOrg].schema
+    val userLookupschema = Encoders.product[UserLookup].schema
+
+    // Read user, user_org and usr_external identity table intot the memory 
+
+    val userdata = spark.read.format("org.apache.spark.sql.cassandra").schema(userschema).option("keyspace", "sunbird").option("table", "user").load();
+    val userOrgData = spark.read.format("org.apache.spark.sql.cassandra").schema(userOrgSchema).option("keyspace","sunbird").option("table","user_org").load();
+    val stateUserExternalIdData = spark.read.format("org.apache.spark.sql.cassandra").schema(stateUserSchema).option("keyspace","sunbird").option("table","usr_external_identity").load();
+
+    // Filter out the user records where all of email, phone, username are null.
```
Comment (indrajra) [inline etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala:70]: can we have the count of records just after loading as well?

File: etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    import spark.implicits._
+    val userOrgMap = userOrgData.select(col("userid"),col("organisationid")).map(r=>(r.getAs[String](0),r.getAs[String](1))).collect.toMap;
+    val translationMap: Column = typedLit(userOrgMap)
+    val userLookupSchema = Encoders.product[UserLookup].schema;
+    val userLookupDF =  spark.read.format("org.apache.spark.sql.cassandra").schema(userLookupSchema).option("keyspace", "sunbird").option("table", "user_lookup").load();
+    
+
+    //Update value as email@orgId, phone@orgId, username@orgId, externalid@orgId
+
+    val stateUserLookupDF = filteredStateUserData.select(col("externalId"),col("userid"),col("provider")).withColumn("type",lit("externalid")).withColumn("value",concat(col("externalid"),lit("@"),col("provider"))).withColumn("userid",col("userid")).select(col("type"),col("value"),col("userid"));
```
Comment (indrajra) [inline etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala:92]: wouldn't this be 123@<provider> ? We wanted to be 123@<orgId>.

File: etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    //Save reecords to the user_lookup table
+
+    stateUserLookupDF.write.format("org.apache.spark.sql.cassandra").option("keyspace", "sunbird").option("table", "user_lookup").mode(SaveMode.Append).save();
+    emailUserLookupDFwithoutNull.write.format("org.apache.spark.sql.cassandra").option("keyspace", "sunbird").option("table", "user_lookup").mode(SaveMode.Append).save();
+    phoneUserLookupDFwithoutNull.write.format("org.apache.spark.sql.cassandra").option("keyspace", "sunbird").option("table", "user_lookup").mode(SaveMode.Append).save();
+    usernameUserLookupDFwithoutNull.write.format("org.apache.spark.sql.cassandra").option("keyspace", "sunbird").option("table", "user_lookup").mode(SaveMode.Append).save();
+
+    val newTableRecords = spark.read.format("org.apache.spark.sql.cassandra").schema(userLookupschema).option("keyspace", "sunbird").option("table", "user_lookup").load().count();
+    println("user_lookup count post migration: " + newTableRecords);
```
Comment (indrajra) [inline etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala:118]: After doing this, I'd like some extra metrics as well to be done - all in the user table (not in lookup).

How many duplicate? If possible do a join on these dataframes and output their uuids.
1. usernames 
2. emails
3. phone

**Review by indrajra — CHANGES_REQUESTED** [review]

File: etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala (hunk 2 above)
Comment (sknirmalkar89) [reply etl-jobs/src/main/scala/org/learner/jobs/UserLookupMigration.scala:92]: it will be 123@orgiId

**Review by indrajra — APPROVED** [review]

**Pull request merged**